channeldb/invoices.go

```diff
@@ -1914,7 +1925,25 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, invoices,
 	// the process by updating the state transitions for individual HTLCs
 	// and recalculate the total amount paid to the invoice.
 	var amtPaid lnwire.MilliSatoshi
-	for _, htlc := range invoice.Htlcs {
+	for key, htlc := range invoice.Htlcs {
+		// Set the HTLC preimage for any AMP HTLCs.
+		if setID != nil {
+			preimage, ok := update.State.HTLCPreimages[key]
+			switch {
+
+			// If we don't already have a preiamge for this HTLC, we
+			// can set it now.
+			case ok && htlc.AMP.Preimage == nil:
+				htlc.AMP.Preimage = &preimage
+
+			// Otherwise, prevent over-writing an existing preimage.
+			// Ignore the case where the preimage is identical.
+			case ok && *htlc.AMP.Preimage != preimage:
+				return nil, ErrHTLCPreimageAlreadyExists
```

Comment:
i would need to check (part of the TODO w/ adding more tests around invalid HTLC sets). as far as probing though, repeating the same share on a different HTLC won't hit this check since it's keyed by CircuitKey. we really shouldn't hit this check ever anyways, it is more of a sanity check to prevent us from making incorrect transitions at the db level.